R/drive_ls.R

```diff
@@ -26,51 +28,161 @@
 #' }
 #'
 #' @export
-drive_ls <- function(search = NULL, ..., fixed = FALSE, verbose = TRUE){
+drive_ls <- function(path = NULL,
+                     pattern = NULL,
+                     ...,
+                     fixed = FALSE,
+                     verbose = TRUE){
+
+  folder <- NULL
 
-  request <- build_drive_ls(...)
+  if (!is.null(path)){
+    folder <- climb_folders(path = path)
+  }
+
+  request <- build_drive_ls(..., folder = folder)
   response <- make_request(request)
-  process_drive_ls(response = response, search = search, fixed = fixed, verbose = verbose)
+  process_drive_ls(
+    response = response,
+    pattern = pattern,
+    fixed = fixed,
+    verbose = verbose
+  )
+
+}
+
+climb_folders <- function(path = NULL, my_folders = drive_folders()){
+
+  folders <- unlist(strsplit(path, "/"))
+  my_folder_ids <- purrr::map(folders, folder_ids, folder_tbl = my_folders)
+
+  if (!any(my_folder_ids[[1]]$root)){
+    spf("We could not find a folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
 
+  #it is v silly that Google Drive allows this but...
+  if (!sum(my_folder_ids[[1]]$root) == 1){
+    spf("It seems you have more than one folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
+
+  if (length(my_folder_ids) == 1){
+    return(my_folder_ids[[1]]$id[my_folder_ids[[1]]$root == TRUE])
+  }
+  len <- length(my_folder_ids)
+
+  while (nrow(my_folder_ids[[len - 1]]) != 1){
+    len <- length(my_folder_ids)
+    f1 <- my_folder_ids[1:(len - 1)]
+    f2 <- my_folder_ids[2:len]
+    my_folder_ids <- purrr::map2(f1, f2, folder_check)
+  }
+  my_folder_ids[[len - 1]]$id
```

Comment:
Oh hmm on my end I need `len-1` because I calculate `len` as the length of the list prior to performing the `map2` function, and after that function the length should be `len-1`, so to grab the last one it has to be `len-1`.